rclcpp/src/rclcpp/executors/executor_entities_collector.cpp

```diff
@@ -0,0 +1,407 @@
+  }
+
+  std::lock_guard<std::mutex> pending_lock(pending_mutex_);
+
+  for (auto weak_node_ptr : pending_added_nodes_) {
+    auto node_ptr = weak_node_ptr.lock();
+    if (node_ptr) {
+      node_ptr->get_associated_with_executor_atomic().store(false);
+    }
+  }
+  pending_added_nodes_.clear();
+  pending_removed_nodes_.clear();
+
+  for (auto weak_group_ptr : pending_manually_added_groups_) {
+    auto group_ptr = weak_group_ptr.lock();
+    if (group_ptr) {
+      group_ptr->get_associated_with_executor_atomic().store(false);
+    }
+  }
+  pending_manually_added_groups_.clear();
+  pending_manually_removed_groups_.clear();
+}
+
+bool ExecutorEntitiesCollector::has_pending()
+{
+  std::lock_guard<std::mutex> pending_lock(pending_mutex_);
+  return pending_manually_added_groups_.size() != 0 ||
+         pending_manually_removed_groups_.size() != 0 ||
+         pending_added_nodes_.size() != 0 ||
+         pending_removed_nodes_.size() != 0;
+}
+
+void
+ExecutorEntitiesCollector::add_node(rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_ptr)
+{
+  // If the node already has an executor
+  std::atomic_bool & has_executor = node_ptr->get_associated_with_executor_atomic();
+  if (has_executor.exchange(true)) {
+    throw std::runtime_error(
+            std::string("Node '") + node_ptr->get_fully_qualified_name() +
+            "' has already been added to an executor.");
+  }
+
+  std::lock_guard<std::mutex> pending_lock(pending_mutex_);
+  bool associated = weak_nodes_.count(node_ptr) != 0;
+  bool add_queued = pending_added_nodes_.count(node_ptr) != 0;
+  bool remove_queued = pending_removed_nodes_.count(node_ptr) != 0;
+
+  if ((associated || add_queued) && !remove_queued) {
+    throw std::runtime_error(
+            std::string("Node '") + node_ptr->get_fully_qualified_name() +
+            "' has already been added to this executor.");
+  }
+
+  this->pending_added_nodes_.insert(node_ptr);
+}
+
+void
+ExecutorEntitiesCollector::remove_node(
+  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_ptr)
+{
+  if (!node_ptr->get_associated_with_executor_atomic().load()) {
+    throw std::runtime_error("Node needs to be associated with an executor.");
+  }
+
+  std::lock_guard<std::mutex> pending_lock(pending_mutex_);
+  bool associated = weak_nodes_.count(node_ptr) != 0;
+  bool add_queued = pending_added_nodes_.count(node_ptr) != 0;
+  bool remove_queued = pending_removed_nodes_.count(node_ptr) != 0;
+
+  if (!(associated || add_queued) || remove_queued) {
+    throw std::runtime_error(
+            std::string("Node '") + node_ptr->get_fully_qualified_name() +
+            "' needs to be associated with this executor.");
+  }
+
+  this->pending_removed_nodes_.insert(node_ptr);
+}
+
+void
+ExecutorEntitiesCollector::add_callback_group(rclcpp::CallbackGroup::SharedPtr group_ptr)
+{
+  std::atomic_bool & has_executor = group_ptr->get_associated_with_executor_atomic();
+  if (has_executor.exchange(true)) {
```

Comment:
@mjcarroll I think that this is a bug.
when you call `add_callback_group` you immediately set this to "true", but then you also add the group to the `pending_manually_added_groups_`.

If you later call `update_collections`, it will try to add again the callback group to the executor and fail, becuase the bool variable was already set.

Reply:
The same is probably true also in the `add_node` function.

I think that it's probably correct to do the "exchange" here, but then we shouldn't do it again in `add_callback_group_to_collection`

Reply:
Fixed in 1b1a915